File: src/webengine/libqwebengineloadinginfo.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qwebengineloadinginfo_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qwebengineloadinginfo.html
pub const qwebengineloadinginfo = struct {
    /// New constructs a new QWebEngineLoadingInfo object.
    ///
    /// ``` other: QtC.QWebEngineLoadingInfo ```
    pub fn New(other: ?*anyopaque) QtC.QWebEngineLoadingInfo {
        return qtc.QWebEngineLoadingInfo_new(@ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#operator-eq)
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo, other: QtC.QWebEngineLoadingInfo ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QWebEngineLoadingInfo_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#url)
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo ```
    pub fn Url(self: ?*anyopaque) QtC.QUrl {
        return qtc.QWebEngineLoadingInfo_Url(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#isErrorPage)
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo ```
    pub fn IsErrorPage(self: ?*anyopaque) bool {
        return qtc.QWebEngineLoadingInfo_IsErrorPage(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#status)
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo ```
    pub fn Status(self: ?*anyopaque) i64 {
        return qtc.QWebEngineLoadingInfo_Status(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#errorString)
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo, allocator: std.mem.Allocator ```
    pub fn ErrorString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebEngineLoadingInfo_ErrorString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebengineloadinginfo.ErrorString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#errorDomain)
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo ```
    pub fn ErrorDomain(self: ?*anyopaque) i64 {
        return qtc.QWebEngineLoadingInfo_ErrorDomain(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#errorCode)
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo ```
    pub fn ErrorCode(self: ?*anyopaque) i32 {
        return qtc.QWebEngineLoadingInfo_ErrorCode(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebengineloadinginfo.html#dtor.QWebEngineLoadingInfo)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QWebEngineLoadingInfo ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWebEngineLoadingInfo_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qwebengineloadinginfo.html#types
pub const enums = struct {
    pub const LoadStatus = enum {
        pub const LoadStartedStatus: i32 = 0;
        pub const LoadStoppedStatus: i32 = 1;
        pub const LoadSucceededStatus: i32 = 2;
        pub const LoadFailedStatus: i32 = 3;
    };

    pub const ErrorDomain = enum {
        pub const NoErrorDomain: i32 = 0;
        pub const InternalErrorDomain: i32 = 1;
        pub const ConnectionErrorDomain: i32 = 2;
        pub const CertificateErrorDomain: i32 = 3;
        pub const HttpErrorDomain: i32 = 4;
        pub const FtpErrorDomain: i32 = 5;
        pub const DnsErrorDomain: i32 = 6;
        pub const HttpStatusCodeDomain: i32 = 7;
    };
};

File: src/webengine/libqwebengineloadinginfo.cpp
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <QWebEngineLoadingInfo>
#include <qwebengineloadinginfo.h>
#include "libqwebengineloadinginfo.h"
#include "libqwebengineloadinginfo.hxx"

QWebEngineLoadingInfo* QWebEngineLoadingInfo_new(const QWebEngineLoadingInfo* other) {
    return new QWebEngineLoadingInfo(*other);
}

void QWebEngineLoadingInfo_OperatorAssign(QWebEngineLoadingInfo* self, const QWebEngineLoadingInfo* other) {
    self->operator=(*other);
}

QUrl* QWebEngineLoadingInfo_Url(const QWebEngineLoadingInfo* self) {
    return new QUrl(self->url());
}

bool QWebEngineLoadingInfo_IsErrorPage(const QWebEngineLoadingInfo* self) {
    return self->isErrorPage();
}

int QWebEngineLoadingInfo_Status(const QWebEngineLoadingInfo* self) {
    return static_cast<int>(self->status());
}

libqt_string QWebEngineLoadingInfo_ErrorString(const QWebEngineLoadingInfo* self) {
    QString _ret = self->errorString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

int QWebEngineLoadingInfo_ErrorDomain(const QWebEngineLoadingInfo* self) {
    return static_cast<int>(self->errorDomain());
}

int QWebEngineLoadingInfo_ErrorCode(const QWebEngineLoadingInfo* self) {
    return self->errorCode();
}

void QWebEngineLoadingInfo_Delete(QWebEngineLoadingInfo* self) {
    delete self;
}
